src/solver/diagonal/forward_diag.cpp

```diff
@@ -0,0 +1,171 @@
+    auto kernel       = KernelInfo{};
+
+    const auto build_params = KernelBuildParameters{
+        {"MIOPEN_USE_FP16", static_cast<int>(dtype == miopenHalf)},
+        {"MIOPEN_USE_FP32", static_cast<int>(dtype == miopenFloat)},
+        {"MIOPEN_USE_FP64", static_cast<int>(dtype == miopenDouble)},
+        {"MIOPEN_USE_BFP16", static_cast<int>(dtype == miopenBFloat16)},
+        {"INPUT_TYPE", input_dtype == "bfloat16" ? "ushort" : input_dtype},
+        {"OUTPUT_TYPE", output_dtype == "bfloat16" ? "ushort" : output_dtype}};
+
+    kernel.comp_options = build_params.GenerateFor(kbp::HIP{});
+
+    if(problem.GetInputDesc().GetLengths().size() == 1)
+    {
+        auto input_numel = problem.GetInputDesc().GetElementSize();
+
+        size_t xlocalsize = LOCAL_SIZE;
+        size_t xgridsize  = AlignUp(input_numel, xlocalsize);
+        size_t ylocalsize = 1;
+        size_t ygridsize  = 1;
+        size_t zlocalsize = 1;
+        size_t zgridsize  = 1;
+
+        kernel.kernel_file = "MIOpenDiag.cpp";
+        kernel.kernel_name = "Diag1dForward";
+
+        kernel.l_wk.push_back(xlocalsize);
+        kernel.l_wk.push_back(ylocalsize);
+        kernel.l_wk.push_back(zlocalsize);
+
+        kernel.g_wk.push_back(xgridsize);
+        kernel.g_wk.push_back(ygridsize);
+        kernel.g_wk.push_back(zgridsize);
+
+        result.construction_params.push_back(kernel);
+
+        result.invoker_factory = [](const std::vector<Kernel>& kernels) {
+            return [=](const Handle& handle_, const AnyInvokeParams& raw_params) {
+                decltype(auto) kernel = handle_.Run(kernels.front());
+                decltype(auto) params =
+                    raw_params.CastTo<miopen::diagonal::diag::FwdInvokeParams>();
+                auto input_numel = params.inputDesc->GetElementSize();
+                auto output_tv   = get_inner_expanded_tv<2>(*params.outputDesc);
+                long offset      = (params.diagonal >= 0 ? params.diagonal * output_tv.stride[1]
+                                                         : -params.diagonal * output_tv.stride[0]);
+
+                kernel(params.input, params.output, input_numel, offset, output_tv);
+            };
+        };
+    }
+    else if(problem.GetInputDesc().GetLengths().size() == 2)
+    {
+        int64_t output_numel = problem.GetOutputDesc().GetElementSize();
+
+        size_t xlocalsize = LOCAL_SIZE;
+        size_t xgridsize  = AlignUp(output_numel, xlocalsize);
+        size_t ylocalsize = 1;
+        size_t ygridsize  = 1;
+        size_t zlocalsize = 1;
+        size_t zgridsize  = 1;
+
+        kernel.kernel_file = "MIOpenDiag.cpp";
+        kernel.kernel_name = "Diag2dForward";
+
+        kernel.l_wk.push_back(xlocalsize);
+        kernel.l_wk.push_back(ylocalsize);
+        kernel.l_wk.push_back(zlocalsize);
+
+        kernel.g_wk.push_back(xgridsize);
+        kernel.g_wk.push_back(ygridsize);
+        kernel.g_wk.push_back(zgridsize);
```

Comment:
There are a lot of duplicate codes in the 2 blocks of code: `if(problem.GetInputDesc().GetLengths().size() == 2)`, `if(problem.GetInputDesc().GetLengths().size() == 1)`.  I see only 2 different lines: `int64_t output_numel = problem.GetOutputDesc().GetElementSize();` and `kernel.kernel_name = "Diag2dForward";`. Plz try to refactor them.